File: Week_06/Practicum/Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def bfs(root):
    q = deque([root])
    
    ans = []

    while q:
        # Each iteration here holds exactly one level of the tree
        ans.append([node.val for node in q])
        for _ in range(len(q)):
            current = q.popleft()
        
            if current.left:
                q.append(current.left)
            if current.right:
                q.append(current.right)
    
    return ans

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        return bfs(root) if root else []

File: Week_06/Practicum/Binary Tree Level Order Traversal.cpp
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) {
            return {};
        }

        std::vector<std::vector<int>> levelOrder;
        std::queue<TreeNode*> q;
        q.push(root);
        
        while (q.size()) {
            std::vector<int> level(q.size());
            for (int i = 0; i < level.size(); ++i) {
                TreeNode* node = q.front(); q.pop();
                level[i] = node->val;
                
                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
            }

            levelOrder.push_back(level);
        }

        return levelOrder;
    }
};
